src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3843,7 +3843,7 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
 
             if ((object)optRightType != null && optRightType.IsDynamic())
             {
-                var leftPlaceholder = new BoundValuePlaceholder(leftOperand.Syntax, optLeftType).MakeCompilerGenerated();
+                var leftPlaceholder = new BoundValuePlaceholder(leftOperand.Syntax, optLeftType0).MakeCompilerGenerated();
```

Comment:
:memo: When the LHS of null-coalescing operator is a nullable value type, in the branch where LHS is not null, the lowering does `.GetValueOrDefault` before converting to the result type. See `GetConvertedLeftForNullCoalescingOperator`.
In the case of `dynamic` as result type, this mean we'll do `(object)left.GetValueOrDefault`, so the conversion here needs to be `leftType0 -> object`.